src/Symfony/Component/Messenger/Transport/Dsn.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport;
+
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+
+/**
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+final class Dsn
+{
+    private $scheme;
+    private $host;
+    private $user;
+    private $password;
+    private $port;
+    private $path;
+    private $options;
+
+    public function __construct(string $scheme, string $host, ?string $user = null, ?string $password = null, ?int $port = null, ?string $path = null, array $options = [])
+    {
+        $this->scheme = $scheme;
+        $this->host = $host;
+        $this->user = $user;
+        $this->password = $password;
+        $this->port = $port;
+        $this->path = $path;
+        $this->options = $options;
+    }
+
+    public static function fromString(string $dsn, array $options): self
+    {
+        if (false === $parsedDsn = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The "%s" messenger DSN is invalid.', $dsn));
+        }
+
+        $user = urldecode($parsedDsn['user'] ?? null);
+        $password = urldecode($parsedDsn['pass'] ?? null);
+        $port = $parsedDsn['port'] ?? null;
+        parse_str($parsedDsn['query'] ?? '', $options);
+
+        $path = $parsedDsn['path'] ?? null;
+
+        return new self($parsedDsn['scheme'], $parsedDsn['host'], $user, $password, $port, $path, $options);
+    }
+
+    public function getScheme(): string
+    {
+        return $this->scheme;
+    }
+
+    public function getHost(string $default = null): string
```

Comment:
The return type is wrong. If there is no host in the DSN and you don't provide a non-null default, it will return `null`